core/state/statedb.go

```diff
@@ -53,6 +53,19 @@ func (n *proofList) Delete(key []byte) error {
 	panic("not supported")
 }
 
+// StateLogger is used to collect state update traces from  EVM transaction
+// execution.
+// Note that reference types are actual VM data structures; make copies
+// if you need to retain them beyond the current call.
```

Comment:
Please also document that these all are invoked post-execution, and the values reported via these methods are the finalized values (as opposed to runtime values which might revert)

Reply:
> Please also document that these all are invoked post-execution

Good tip

> these methods are the finalized values (as opposed to runtime values which might revert)

Is that the case? I thought values in statedb could still revert.

Reply:
Right, my bad

Reply:
SO, how does a "tracer" (generically, I know there differen types), track

1. A transfer in some subscope, and the subscope reverts five ops later
2. A transfer in some subscope, and the grand-parent scope reverts, 100 ops later
3. A transfer in some subscope, but the transaction goes OOG and reverts later on
4. A transfer in some subscope, but the block `A` is not yet finalized. The same tx is also executed on a sibling block `B`, and the transfer is different there. The same tx is also executed on sibling block `C`, and there the transfer never happens.

Reply:
I'm unsure about some of your terminology. I'll assume `subscope` means a new `CALL/STATIC/DELEGATE/etc.`  execution context and `ops` means opcode executed.

> A transfer in some subscope, and the subscope reverts five ops later

```
- Trx (also root call)
-   CALL [pure transfer] -> Fails
```

Then `CaptureExit` will have the `err` to determine the call failed, reverted/failed can be checked using `error.Is(err, evm.ExecutionReverted)`.

> A transfer in some subscope, and the grand-parent scope reverts, 100 ops later
> A transfer in some subscope, but the transaction goes OOG and reverts later on

```
- Trx (also root call)
-   CALL (contract's call)
-       CALL (another one)
-           CALL [pure transfer]
-           <-- Failure Point -->
```

In those cases, when the transaction terminates, in `CaptureTxEnd`, we re-create the call tree structure of the trx and if a parent node is reverted/failed, we flag all the calls in its substree with a special value to tell that this call did not make it to the chain.

> A transfer in some subscope, but the block A is not yet finalized. The same tx is also executed on a sibling block B, and the transfer is different there. The same tx is also executed on sibling block C, and there the transfer never happens.

I think in those it depends on the use case(s) of how the tracer is going to be used. Some tracers could wait until the block is actually finalized (there is implementation challenges for this but it's feasible. In our case, we deal with fork in a later step. So we receive all blocks from the tracer, then we deal with forks after the tracer by keeping the parent/child relationship in a "tree".